### PR TITLE
Introduce Task and AdminComponent

### PR DESCRIPTION
**Pull request by hakonhall**
- Make an optional include of variant.xml in services.xml to allow custom node admin components.
 - Defines Task which is similar to a resource/recipe in Chef. A Task can invoke subtasks so it forms a tree. A task is supposed to be executed now and without delay.
 - Defines an AdminComponent which must be explicitly enabled, and moves the current node admin bulk within a DockerAdminComponent. AdminComponent needs to be refined in later PRs.

### CODE REVIEW

**Comment by hakonhall** [comment]
@freva , could you take a brief look at the PR to ensure we agree on the general direction this is taking? I'll add unit tests once that has been done.

FYI @smorgrav

**Comment by hakonhall** [comment]
Btw @smorgrav , I'm adding NodeAdminConfig/DockerAdminConfig to DockerAdminComponent to allow you to add stuff only for aws (aws_tenant mode).

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTask.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    public WriteFileTask(Params params) {
+        this.params = params;
+    }
+
+    @Override
+    public boolean execute(TaskContext context) {
+        final FileSystemPath path = context.getFileSystem().withPath(params.path);
+        if (path.isRegularFile()) {
```
Comment (freva) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTask.java]: Should maybe compare file contents? I would expect `WriteFileTask` to overwrite an old file should the contents differ.

**Review by freva — APPROVED** [review]

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTask.java (hunk 1 above)
Comment (hakonhall) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTask.java]: Let me add a TODO for that. Some file content may be expensive to compute compared to an simpler check like existence(or timestamp, so perhaps this needs to be controlled by parameters.

Comment (hakonhall) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTask.java]: Actually there is a TODO below, but let me move it up here and rephrase.

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/io/FileSystem.java [hunk 2]
```diff
@@ -36,26 +37,63 @@ public void createDirectory(Path path, FileAttribute<?>... attributes) {
...
+    private PosixFileAttributes getAttributes(Path path) {
+        return uncheck(() ->
+                Files.getFileAttributeView(path, PosixFileAttributeView.class).readAttributes());
+    }
+
+    public Set<PosixFilePermission> getPermissions(Path path) {
+        return getAttributes(path).permissions();
+    }
+
+    static void validatePermissions(String permissions) {
```
Comment (freva) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/io/FileSystem.java]: This method is not used?

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/io/FileSystem.java [hunk 3]
```diff
@@ -36,26 +37,63 @@ public void createDirectory(Path path, FileAttribute<?>... attributes) {
...
         byte[] contentInUtf8 = content.getBytes(StandardCharsets.UTF_8);
         uncheck(() -> Files.write(path, contentInUtf8, options));
     }
 
+    private PosixFileAttributes getAttributes(Path path) {
+        return uncheck(() ->
+                Files.getFileAttributeView(path, PosixFileAttributeView.class).readAttributes());
+    }
+
+    public Set<PosixFilePermission> getPermissions(Path path) {
```
Comment (freva) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/io/FileSystem.java]: Consider returning `String` so it matches the set counterpart and simplifies testing.

**Comment by hakonhall** [comment]
Please take another look @freva

File: node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTaskTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+public class WriteFileTaskTest extends TaskTestBase {
+    private final String content = "line1\nline2\n";
+
+    @Test
+    public void testWrite() {
+        Path path = Paths.get("foo");
+
+        @SuppressWarnings("unchecked")
+        Producer<String> contentProducer = (Producer<String>) mock(Producer.class);
```
Comment (freva) [inline node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTaskTest.java]: Why not just `Producer<String> contentProducer = () -> content;` ?

File: node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTaskTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+        assertTrue(task.execute(contextMock));
+
+        verify(fileSystemMock).writeUtf8File(path, content);
+        verify(fileSystemMock).setPermissions(path, permissions);
+        verify(fileSystemMock).setOwner(path, owner);
+        verify(fileSystemMock).setGroup(path, group);
+
+        // Writing a file with the expected content
+        ArgumentCaptor<WriteFileTask> writeFileTaskArgumentCaptor =
```
Comment (freva) [inline node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTaskTest.java]: What is this for? I don't see it being used

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/io/FileSystem.java (hunk 3 above)
Comment (hakonhall) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/io/FileSystem.java]: Done

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/io/FileSystem.java (hunk 2 above)
Comment (hakonhall) [reply node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/io/FileSystem.java]: Removed

File: node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTaskTest.java (hunk 4 above)
Comment (hakonhall) [reply node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTaskTest.java]: I wanted to verify it was being called once: Added

File: node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTaskTest.java (hunk 5 above)
Comment (hakonhall) [reply node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/task/WriteFileTaskTest.java]: Added verification of directory creation subtask.

**Review by freva — APPROVED** [review]

**Pull request merged**